rfd/0156-notifications.md

```diff
@@ -0,0 +1,569 @@
+---
+authors: Yassine Bounekhla (yassine@goteleport.com)
+state: draft
+---
+
+# RFD 0156 - Notifications
+
+## Required Approvals
+
+- Engineering: @zmb3
+- Product: @roraback
+
+## What
+
+A notification system that can be used to notify users of various events, updates, and warnings in real time.
+
+## Why
+
+Users currently have no way to easily view or be notified of updates and events relevant to them. Our existing cluster alerts system is cumbersome and doesn't support many
+desired use cases due to its non-dynamic nature. With the notification system, features in the application will be able to programmatically send notifications to specific user(s) to
+notify them of relevant events. Cluster admins will also be able to manually send informational notifications to their cluster's users.
+
+Examples of potential use cases include being able to notify users of:
+
+- A new access request that needs their review.
+- Their access request being approved/denied.
+- Needing to complete their periodic access list review.
+- An access monitoring alert.
+- A failed payment.
+- Approaching a usage limit.
+- Needing to upgrade their Teleport version.
+- New features added to Teleport.
+
+## Details
+
+Notifications will be stored in the backend and be of two types:
+
+1. User-specific notifications: Targets a single specific user. An example is a notification to notify a user that their access request has been approved.
+2. Global notifications: Targets multiple users. An example is a notification to notify users that there is a new access request needing review.
+   They can be configured to target users based on one of the following matchers:
+   1. Permissions: Targets all users with the specified RBAC permission(s).
+   2. Role: Targets all users with the specified role(s).
+   3. All: Targets all users.
+
+Since user-specific notifications are stored by username, notifications for SSO users will persist between sessions.
+
+Notifications will have a default TTL of 30 days, after which they will be automatically deleted. In some cases, notifications may also automatically be deleted prior to their TTL if configured to do so.
+
+#### Backend
+
+User-specific notifications will be stored in the backend under the key `/notifications/user/<auth connector>/<the user's username>/<notification id>`.
+
+We include the `auth connector` in the key to prevent the edge case where a SSO user has the same username as a local user (or a user using another auth connector), this would cause the users' notifications to be merged.
```

Comment:
I don't have any strong opinion either way, I believe the concern was more UX related as it could be quite confusing for a user if this edge case does occur. What do you think? @zmb3 

> Access-control for alerts shouldn't be static anyway. User roles change over time, so whether or not viewing a given notification is still allowable needs to be reevaluated when it is loaded anyway.

This is for user-specific notifications, which don't involve any checking for access controls. The notifications that match based on roles and permissions (ie. global notifications) are dynamic and are evaluated on load.